doc/design/v2_layer.md

```diff
@@ -0,0 +1,126 @@
+# Using tape to refactor current paddle.v2 configuration parsing
+
+## What are the problems
+
+Paddle V2 API give a flexible way to configure neural network topology. The user can create a neural network topology layer by layer. We use the final layer to represent the neural network topology.
+
+The example code for users is:
+
+```python
+img = data_layer(name="img", type=Dense(784))
+hidden = fc_layer(input=img, size=200)
+prediction = fc_layer(input=hidden, size=10, act=Softmax())
+cost = classification_cost(input=prediction, label=data_layer(name="label", type=Integer(10)))
+
+paddle.train(cost, ...)
+```
+
+We use `cost` to represent the entire topology of the neural network.  We use the last node to traverse the entire topology by `depth first search` algorithm. The connection between each layer is represented by the `input` parameter.  It is fit for representing a plain neural network topology. However, there are some special layers in Paddle, which are not connected explicitly by `input` parameter. They are:
+
+* Evaluator.  An evaluator is used to compute metrics(such as error rate, f1 score) in Paddle. An evaluator can not be the input of other layers. So we cannot access evaluators by simply traversing back from the final layer.
+* Memory. We use memory layers in Recurrent Neural Network; a memory layer represents the output of some layer in the last time step. However, the memory layer connects to its input layer implicitly by sharing the same name. We also cannot traverse to memory layer because maybe there is no layer using this memory layer.
+* Recurrent Group.  The recurrent group is a sub-topology config using in Recurrent Neural Network. It represents the layers in each recurrent time-step. We could traverse back to some topology in a recurrent group, but the sub-topology is non-splittable. The recurrent group should be either entirely in the topology or not in the topology. 
```

Comment:
I just realized that I am not familiar with that how PaddlePaddle represent RNN yet. Any suggestion on how I can get familiar with it so could I help review this design doc? @reyoung